File: canonicalize.pl
#!/user/bin/perl
use strict;
use warnings;
use List::MoreUtils qw/any/;

our @PRIMES = (undef,2,3,5,7,11,13,17,19,23,29,31,41,43,53,59,61,67,71);

sub canonicalize {
    my $atoms = shift;
    my $checked_stereochemistry = 0;
    set_initial_rank( $atoms );
    $atoms = [sort compare_rank @$atoms];
    reduce( $atoms );
    
    while ( $atoms->[-1]->{rank} < @$atoms ) {
        apply_ranks( $atoms );    
        set_primes( $atoms );
        compute_extended_connectivity( $atoms );
        sort_by_rank( $atoms );
        reduce( $atoms );
        
        if( !rank_has_changed( $atoms ) ) {
            if ( !$checked_stereochemistry ) {
                identify_chirals( $atoms );
                identify_isomers( $atoms );
                $checked_stereochemistry = 1;
            }
            break_ties( $atoms );
            reduce( $atoms );
        }
    }
    
    identify_rings( $atoms->[0], 1 );
    reset_search_flags( $atoms );
    print stringify( $atoms->[0] ) . "\n";
}

sub set_initial_rank {
    my $atoms = shift;
    
    foreach my $atom ( @$atoms ) {
        my $atomic_number = $atom->{number};
        my $connections   = $atom->{connections};
        my $non_h_bonds   = $atom->{non_h_bonds};    
        my $sign          = ($atom->{charge} > 0) ? 1 : 0;
        my $charge        = $atom->{charge};
        my $hydrogens     = $atom->{hydrogens};
        $atom->{rank}     = $connections . $non_h_bonds . $atomic_number . $sign . $charge . $hydrogens;
    }
    
    return $atoms;
}

sub sort_by_rank {
    my $atoms    = shift;
    my @sorted   = ();
    my @subset   = ();
    my $previous = -1;
    
    for( my $i = 0; $i < @$atoms; $i++ ) {
        if ( $atoms->[$i]->{previous_rank} != $previous ) {
            $previous = $atoms->[$i]->{previous_rank};
            if ( scalar @subset ) {
                push @sorted, sort compare_rank @subset;
            }
            @subset = ();
        }
        
        push @subset, $atoms->[$i];
    }
    
    if ( scalar @subset ) {
        push @sorted, sort compare_rank @subset;        
    }
    
    @$atoms = @sorted;
    
    return $atoms;
}

sub reduce {
    my $atoms    = shift;
    my $previous = -1;
    my $rank     = -1;
    my $reduced  = 0;
    for( my $i = 0; $i < @$atoms; $i++ ) {
        if ( ($atoms->[$i]->{previous_rank} && $atoms->[$i]->{previous_rank} != $previous) || $atoms->[$i]->{rank} != $rank ) {
            $previous = $atoms->[$i]->{previous_rank};
            $rank     = $atoms->[$i]->{rank};
            $reduced++;
        }
        $atoms->[$i]->{rank} = $reduced;
    }
    
    my @ordered = sort { $a->{id} <=> $b->{id} } @$atoms;
    my @ranks   = map { $_->{rank} } @ordered;
    print join(',', @ranks) . "\n";
    
    return $atoms;
}

sub set_primes {
    my $atoms = shift;
    foreach my $atom ( @$atoms ) {
        $atom->{prime} = $PRIMES[$atom->{rank}];
    }
    
    return $atoms
}

sub compute_extended_connectivity {
    my $atoms = shift;
    foreach my $atom( @$atoms ) {
        my $product = 1;
        foreach my $bonded ( @{$atom->{bonds}} ) {
            $product *= $bonded->{prime};
        }
        $atom->{rank} = $product;
    }
    
    return $atoms;    
}

sub rank_has_changed {
    my $atoms = shift;
    my $changed = 0;
    foreach my $atom ( @$atoms ) {
        if ( $atom->{previous_rank} != $atom->{rank} ) {
            $changed = 1;
            last;
        }
    }
    
    return $changed;
}

sub break_ties {
    my $atoms    = shift;
    my $previous = -1;
    my $broken   = 0;
    
    for( my $i = 0; $i < @$atoms; $i++ ) {
        $atoms->[$i]->{rank} *= 2;
        if( !$broken && $atoms->[$i]->{rank} == $previous ) {
            $broken = 1;
            $atoms->[$i-1]->{rank} -= 1;
        }
        $previous = $atoms->[$i]->{rank};
    }
    
    return $atoms;
}

sub apply_ranks {
    my $atoms = shift;
    
    for ( my $i = 0; $i < @$atoms; $i++ ) {
        my $atom = $atoms->[$i];
        $atom->{previous_rank} = $atom->{rank};
    }
    
    return $atoms;
}

sub compare_rank {
    return $a->{rank} <=> $b->{rank};
}

sub reset_search_flags {
    my $atoms = shift;
    foreach my $atom ( @$atoms ) {
        $atom->{seen} = 0;
    }
    
    return;
}

sub identify_chirals {
    my $atoms = shift;
    foreach my $atom ( @$atoms ) {
        # must be a carbon
        next unless $atom->{number} == 6;
        # must have 4 bonded atoms
        next unless scalar @{$atom->{bonds}} == 4 || scalar @{$atom->{bonds}} == 3 && $atom->{hydrogens} == 1;
        my $rankings = {};
        for my $bonded ( @{$atom->{bonds}} ) {
            $rankings->{$bonded->{rank}} = 1;
        }
        # all bonded atoms ( or groups of atoms ) must be different
        next unless scalar keys %{$rankings} == scalar @{$atom->{bonds}};
        # we are chiral!!
        $atom->{chirality} = '@';
    }
    return;
}

sub identify_isomers {
    my $atoms = shift;
    return;
}

sub identify_rings {
    my ($node, $ring_count, $previous) = @_;
    my @sorted = sort compare_rank @{$node->{bonds}};
    $node->{seen} = 1;
    
    for( my $i = 0; $i < @sorted; $i++ ) {
        if ( $previous && $sorted[$i]->{seen} && $sorted[$i]->{id} != $previous->{id} ) {
            if ( !any { $_ == $ring_count} @{$node->{rings}} ) {
                push @{$node->{rings}}, $ring_count;
                push @{$sorted[$i]->{rings}}, $ring_count;
                $ring_count++;
            }
        }
        
        next if $sorted[$i]->{seen};
        
        identify_rings( $sorted[$i], $ring_count, $node );
    }
    
    return;
}

sub stringify {
    my ($node) = @_;
    my @sorted = sort compare_rank @{$node->{bonds}};
    my $smiles = $node->{symbol};
    
    $smiles .= $node->{chirality} if $node->{chirality};
    $smiles .= join ',', @{$node->{rings}} if $node->{rings};
    
    $node->{seen} = 1;
    
    for( my $i = 0; $i < @sorted; $i++ ) {
        next if $sorted[$i]->{seen};
        
        my $bonded_smiles = stringify( $sorted[$i] );
        if ( unseen_count( \@sorted ) ) {
            $bonded_smiles = '(' . $bonded_smiles . ')';
        }
        
        $smiles .= $bonded_smiles;
    }
    
    return $smiles;
}

sub unseen_count {
    my $atoms = shift;
    my $unseen = 0;
    foreach my $atom ( @$atoms ) {
        $unseen += 1 unless $atom->{seen};
    }
    
    return $unseen;
}

File: js/canonicalize.js
var PRIMES = [undefined,2,3,5,7,11,13,17,19,23,29,31,41,43,53,59,61,67,71];

function canonicalize(atoms) {
    var checked_stereochemistry = false;
    
    // rank and sort atoms
    set_initial_rank(atoms);
    atoms.sort(compare_rank);
    reduce(atoms);
    
    while (atoms[atoms.length-1].rank < atoms.length ) {
        apply_ranks(atoms);    
        set_primes(atoms);
        compute_extended_connectivity(atoms);
        sort_by_rank(atoms);
        reduce(atoms);
        
        if (!rank_has_changed(atoms)) {
            if (!checked_stereochemistry) {
                identify_chirals(atoms);
                identify_isomers(atoms);
                checked_stereochemistry = true;
            }
            break_ties(atoms);
            reduce(atoms);
        }
    }
    
    identify_rings(atoms[0], 1);
    reset_search_flags(atoms);

    // print out each string
    console.log(stringify(atoms[0]) + "\n");
}

function set_initial_rank(atoms) {
    for (var atom in atoms) {
        var sign = atom.charge > 0 ? 1 : 0;
        atom.rank = atom.connections 
            + atoms.non_h_bonds 
            + atoms.number 
            + sign 
            + atom.charge 
            + atom.hydrogens;
    }
    
    return atoms;
}

function sort_by_rank(atoms) {
    var sorted   = [],
        subset   = [],
        previous = -1;
    
    for (var i = 0; i < atoms.length; i++) {
        if (atoms[i].previous_rank != previous ) {
            previous = atoms[i].previous_rank;
            if (subset.length) {
                // sort subset and add to sorted
                sorted.push(subset.sort(compare_rank));
            }
            subset = [];
        }
        
        subset.push(atoms[i]);
    }
    
    if (subset.length) {
        sorted.push(subset.sort(compare_rank));       
    }
    
    atoms = sorted;
    return atoms;
}

function reduce(atoms) {
    var previous = -1,
        rank     = -1,
        reduced  = 0;

    for (var i = 0; i < atoms.length; i++) {
        var atom = atoms[i];
        if ((atom.previous_rank && atom.previous_rank != previous) || atom.rank != rank) {
            previous = atom.previous_rank;
            rank     = atom.rank;
            reduced++;
        }

        atom.rank = reduced;
    }
    
    //my @ordered = sort { $a->{id} <=> $b->{id} } @$atoms;
    //my @ranks   = map { $_->{rank} } @ordered;
    //print join(',', @ranks) . "\n";
    
    return atoms;
}

function set_primes(atoms) {
    for (var i = atoms.length; --i >= 0;) {
        var atom = atoms[i];
        atom.prime = PRIMES[atom.rank];
    }
    
    return atoms;
}

function compute_extended_connectivity(atoms) {
    for (var i = atoms.length; --i >= 0;) {
        var atom = atoms[i];
        var product = 1;

        for (var j = atom.bonds.length; --j >= 0;) {
            product *= atom.bonds[j].prime;
        }

        atom.rank = product;
    }
    
    return atoms;    
}

function rank_has_changed(atoms) {
    var changed = false;

    for (var i = atoms.length; --i >= 0;) {
        var atom = atoms[i];
        if ( atom.previous_rank != atom.rank ) {
            changed = true;
            break;
        }
    }
    
    return changed;
}

function break_ties(atoms) {
    var previous = -1,
        broken   = false;
    
    for (var i = 0; i < atoms.length; i++) {
        var atom = atoms[i];
        atom.rank *= 2;

        if (!broken && atom.rank == previous) {
            broken = true;
            atoms[i-1].rank -= 1;
        }

        previous = atom.rank;
    }
    
    return atoms;
}

function apply_ranks(atoms) {
    for (var i = 0; i < atoms.length; i++) {
        var atom = atoms[i];
        atom.previous_rank = atom.rank;
    }
    
    return atoms;
}

function compare_rank(a, b) {
    if (a.rank < b.rank) {
        return -1;
    }
    if (a.rank > b.rank) {
        return 1;
    }

    return 0;
}

function reset_search_flags(atoms) {
    for (var i = atoms.length; --i >= 0;) {
        atoms[i].seen = 0;
    }
    
    return;
}

function identify_chirals(atoms) {
    for (var i = 0; i < atoms.length; i++) {
        var atom = atoms[i];
        // must be a carbon
        if (atom.number != 6) continue;
        // must have 4 bonded atoms
        if (!(atom.bonds.length == 4 || atom.bonds.length == 3 && atom.hydrogens == 1)) continue;
        
        var rankings = {},
            rank_count = 0;

        for (var i = atom.bonds.length; --i >= 0;) {
            var bonded = atom.bonds[i];
            if (!ranking[bonded.rank]) {
                ranking[bonded.rank] = 1;
                rank_count++;
            }
        }
        
        // all bonded atoms ( or groups of atoms ) must be different
        if (rank_count != atom.bonds.length) continue;
        
        // we are chiral!!
        atom.chirality = '@';
    }

    return;
}

function identify_isomers(atoms) {
    return;
}

function identify_rings(node, ring_count, previous) {
    var sorted = node.bonds.sort(compare_rank);

    node.seen = 1;
    
    for (var i = 0; i < sorted.length; i++) {
        if (previous && sorted[i].seen && sorted[i].id != previous.id) {
            if (!has_matching_ring(node.rings, ring_count)) {
                node.rings.push(ring_count);
                sorted[i].rings.push(ring_count);
                ring_count++;
            }
        }
        
        if (sorted[i].seen) continue;
        
        identify_rings(sorted[i], ring_count, node);
    }
    
    return;
}

function has_matching_ring(ring_list, ring_identifier) {
    for (var i = ring_list.length; --i >= 0;) {
        if (ring_list[i] == ring_identifier) {
            return true;
        }
    }

    return false;
}

function stringify(node) {
    var sorted = node.bonds.sort(compare_rank);
    var smiles = node.symbol;
    
    if (node.chirality) smiles += node.chirality;
    if (node.rings)     smiles += node.rings.join(',');
    
    node.seen = 1;
    
    for (var i = 0; i < sorted.length; i++) {
        if (sorted[i].seen) continue;
        
        var bonded_smiles = stringify(sorted[i]);
        if (unseen_count(sorted)) {
            bonded_smiles = '(' + bonded_smiles + ')';
        }
        
        smiles += bonded_smiles;
    }
    
    return smiles;
}

function unseen_count(atoms) {
    var unseen = 0;
    for(var i = atoms.length; --i >= 0; ) {
        if (!atoms[i].seen) {
            unseen += 1;
        }
    }
    
    return unseen;
}